ui/src/TestAzureTreeDataProvider.ts

```diff
@@ -0,0 +1,237 @@
+        if (!Array.isArray(expectedContextValues)) {
+            expectedContextValues = [expectedContextValues];
+        }
+
+        // tslint:disable-next-line:strict-boolean-expressions
+        let node: IAzureNode = startingNode || this._subscriptionNodes[0]; // automatically just grab the first (test) subscription node
+        while (!expectedContextValues.some((val: string) => node.treeItem.contextValue === val)) {
+            if (node instanceof AzureParentNode) {
+                node = await node.pickChildNode(expectedContextValues);
+            } else {
+                throw new Error(localize('noResourcesError', 'No matching resources found.'));
+            }
+        }
+
+        return node;
+    }
+
+    public async findNode(id: string): Promise<IAzureNode | undefined> {
+        let nodes: IAzureNode[] = await this.getChildren();
+        let foundAncestor: boolean;
+
+        do {
+            foundAncestor = false;
+
+            for (const node of nodes) {
+                if (node.id === id) {
+                    return node;
+                } else if (id.startsWith(`${node.id}/`) && node instanceof AzureParentNode) {
+                    // Append '/' to 'node.id' when checking 'startsWith' to ensure its actually an ancestor, rather than a node at the same level that _happens_ to start with the same id
+                    // For example, two databases named 'test' and 'test1' as described in this issue: https://github.com/Microsoft/vscode-cosmosdb/issues/488
+                    nodes = await node.getCachedChildren();
+                    foundAncestor = true;
+                    break;
+                }
+            }
+        } while (foundAncestor);
+
+        return undefined;
+    }
+
+    private async getRootNodes(actionContext: IActionContext): Promise<IAzureNode[]> {
+        actionContext.properties.isActivationEvent = 'true';
+        actionContext.properties.contextValue = 'root';
+        let nodes: IAzureNode[];
+        this._subscriptionNodes = [];
+        type servicePrincipalCredentials = ApplicationTokenCredentials & { environment: AzureEnvironment };
+        const clientId: string | undefined = process.env.SERVICE_PRINCIPAL_CLIENT_ID ;
+        const secret: string | undefined = process.env.SERVICE_PRINCIPAL_SECRET;
+        const domain: string | undefined = process.env.SERVICE_PRINCIPAL_DOMAIN;
+
+        if (!clientId || !secret || !domain) {
+            throw new Error('Tests can only be run on Travis.');
+        }
+        const credentials: servicePrincipalCredentials = <servicePrincipalCredentials>(await loginWithServicePrincipalSecret(clientId, secret, domain));
+        const subscriptionClient: SubscriptionClient = new SubscriptionClient(credentials);
+        const subscriptions: SubscriptionListResult = await subscriptionClient.subscriptions.list();
+        // returns an array withy subscriptionId, displayName
```

Comment:
"withy"?  Is that like "truthy" and "falsey"?

Reply:
:ups